system/protocols/generic/factory.py

```diff
@@ -0,0 +1,241 @@
+        passing *self* in as the factory. For example:
+
+        >>> reactor.connectTCP(
+        ...     self.config["network"]["hostname"],
+        ...     self.config["network"]["port"],
+        ...     self,
+        ...     self.config["network"]["timeout"],
+        ...     bindAddress=("0.0.0.0", 0)
+        ... )
+        >>>
+
+        You **must** override this method, or your factory will not work.
+        """
+        raise NotImplementedError()
+
+    # Custom Ultros functions - Not necessary to override in most cases
+
+    def load_module(self):
+        """
+        Load the protocol module, reloading if it's already loaded.
+
+        If you use the standard Ultros naming scheme, then you won't need to
+        override this. For reference, this function expects your protocol to
+        reside in *system.protocols.[name].protocol*.
+        """
+
+        if "main" not in self.config:
+            raise KeyError(
+                "Protocol configuration is missing a 'main' section"
+            )
+
+        if "protocol-type" not in self.config["main"]:
+            raise KeyError(
+                "Protocol configuration is missing a 'protocol-type' value in "
+                "the 'main' section"
+            )
+
+        protocol_type = self.config["main"]["protocol-type"]
+
+        if self.protocol_module is None:
+            self.logger.debug("First-time setup, not reloading")
+            self.protocol_module = importlib.import_module(
+                "system.protocols.{}.protocol".format(
+                    protocol_type
+                )
+            )
+        else:
+            self.logger.debug("Reloading module")
+            try:
+                self.protocol_module = reload(self.protocol_module)
+            except:
+                self.protocol_module = None
+                raise
+
+        if not issubclass(self.protocol_module.Protocol, Protocol):
+            self.protocol_module = None
+            raise TypeError(
+                "Protocol '{}' does not subclass the generic protocol "
+                "class!".format(
+                    protocol_type
+                )
+            )
+
+    def get_reconnect_config(self):
+        if "reconnections" in self.config:
+            return self.config["reconnections"]
+        return self.reconnection_config
+
+    def maybe_reconnect(self, connector, reason):
+        reconnection_config = self.get_reconnect_config()
+
+        if self.shutting_down:
+            return
+
+        if reconnection_config.get("on-{}".format(reason), False):
+            if (
+                    self.reconnection_attempts >=
+                    reconnection_config["attempts"]
+            ):
+                self.logger.warn(
+                    "Unable to connect after {} attempts, aborting".format(
+                        self.reconnection_attempts
+                    )
+                )
+                return
+
+            self.reconnection_attempts += 1
```

Comment:
Ok, my previous comments were done on the basis that the code from here on were not included in the above if block, which caused the confusion (one was, one wasn't, I didn't notice the second wasn't the same as the first).

My comment on checking if `attempt <= 0` was that setting it to 0 could work as a signal value to not bother reconnecting, making the separate `on-failure` option redundant. I didn't really consider the fact that we have to work with existing configs though, so that change wouldn't make sense. We could however use 0 to mean "no limit", since I don't want to ever stop trying to reconnect. My config currently has it set to some arbitrarily high value instead.